Stokes Cavity - non-newtonian

Jon83Carvalho opened the pull request:
As per e-mail from Mr. Guyer. Uploading jupyter notebook where I incorporated some code to work with Ostwald fluid in the Stokes cavity.

Regards,
Jonas

wd15 commented:
@Jon83Carvalho, thanks for taking the time out to do this. I think that this would be a nice contribution, however, I don't think it should go in the `fipy` directory. Possibly we could put it in something like `notebooks-contributed` or `examples-contributed`.  Are all the png and csv files necessary? Did you mean to include these in the commit?

Reading though it, it seems that you can remove the last cell with `plt.savefig('Stokes3M.eps',format='eps')` in it. If we set up automatic testing for notebooks then we won't want it saving out a file. It looks like it isn't necessary to have all the csv files so removing those from the notebook and from the commit would be good as well.

Do you have any idea how good the solution is? Accuracy?

Thanks for this though. It really is very nice. Just needs some cleaning up.

Here is a list of the things to fix,
 
 - Remove reading and writing files 
 - Few spelling mistakes (jupyter has a plugin for spelling)
 - Use markdown syntax for the refs so that they link
 - Move from `fipy` to `examples-contributed`
 - Only have the .ipynb file in the commit
 - I like the streamlines, but maybe clean up the cells at the end with the repetitive looping

guyer commented:
@Jon83Carvalho I also think this is a welcome contribution. I agree with most of @wd15's recommended changes.

My few amendments are:
- [ ] I think this should just go in `examples/flow`, although I'd call it  `stokesNonNewtonian.ipynb` or similar.
- [ ] Change the viewer to [`MatplotlibStreamViewer`](https://www.ctcms.nist.gov/fipy/fipy/generated/fipy.viewers.matplotlibViewer.html#module-fipy.viewers.matplotlibViewer.matplotlibStreamViewer).
- [ ] I think everything after the sweep loop can go away.

Jon83Carvalho commented:
Thanks for the inputs.

I will comment in details later.

Have a nice weekend.

Regards,
Jonas

On Fri, Jul 31, 2020 at 6:04 PM Jonathan Guyer <notifications@github.com>
wrote:

> @Jon83Carvalho <https://github.com/Jon83Carvalho> I also think this is a
> welcome contribution. I agree with most of @wd15 <https://github.com/wd15>'s
> recommended changes.
>
> My few amendments are:
>
>    - I think this should just go in examples/flow, although I'd call it
>    stokesNonNewtonian.ipynb or similar.
>    - Change the viewer to MatplotlibStreamViewer
>    <https://www.ctcms.nist.gov/fipy/fipy/generated/fipy.viewers.matplotlibViewer.html#module-fipy.viewers.matplotlibViewer.matplotlibStreamViewer>
>    .
>    - I think everything after the sweep loop can go away.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/usnistgov/fipy/pull/748#issuecomment-667355750>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGDDPDRAXGZ6D5DWE7GGGDTR6MWULANCNFSM4PO2K7LQ>
> .
>


-- 
Jonas Carvalho
Petrópolis - RJ
+55 24 988484411
+55 24 22314508

Jon83Carvalho commented:
Will incorporate changes as per review

guyer commented:
There's no need to close this. Just make edits to your notebook and commit them.

Jon83Carvalho commented:
> @Jon83Carvalho, thanks for taking the time out to do this. I think that this would be a nice contribution, however, I don't think it should go in the `fipy` directory. Possibly we could put it in something like `notebooks-contributed` or `examples-contributed`. Are all the png and csv files necessary? Did you mean to include these in the commit?
> 
> Reading though it, it seems that you can remove the last cell with `plt.savefig('Stokes3M.eps',format='eps')` in it. If we set up automatic testing for notebooks then we won't want it saving out a file. It looks like it isn't necessary to have all the csv files so removing those from the notebook and from the commit would be good as well.
> 
> Do you have any idea how good the solution is? Accuracy?
> 
> Thanks for this though. It really is very nice. Just needs some cleaning up.
> 
> Here is a list of the things to fix,
> 
> * Remove reading and writing files
> * Few spelling mistakes (jupyter has a plugin for spelling)
> * Use markdown syntax for the refs so that they link
> * Move from `fipy` to `examples-contributed`
> * Only have the .ipynb file in the commit
> * I like the streamlines, but maybe clean up the cells at the end with the repetitive looping

Hi Mr. Wheeler, thanks for the inputs.

Find some comments and answers below.

Do you have any idea how good the solution is? Accuracy?
I couldn't find an article where it does exactly what is done here, which is using the momentum equation with no terms from the material derivative (which can be considered a viscous dominant flow, very low Reynolds). I tested the algorithm with maximum m=1.5 and minimum 0.5. Made it converge to maximum residual less than 0.01. Paired comparison, of m=1 and the Newtonian solution, using a linear regression yields a coefficient of ~1 meaning that the speeds of the anisotropic model to cover non Newtonian fluids results in speeds nearly the same as in the case where we adopt a Newtonian fluid (the original Stokes cavity example). In addition, the final results with varying values of "m" make sense in terms of the expected fluid behavior for Shear-thinning (m<1) and shear-thickening (m>1). Of course, those tests are not comparable to a Scientific article results and thus, one may think it is worthy to invest more time testing. The non linearity with the "m" term included makes convergence very costly and I haven´t tested many of different values of the relaxation factors.

My intent, when sharing, was also to get help on checking the quality of the result and code improvements suggestions. My future plans are to include Material derivative terms and test in different mesh (Axisymmetric tube flow) and have a Non-newtonian simple solver for some cases I want to investigate my own.

 Below some comment on your questions.
* Few spelling mistakes (jupyter has a plugin for spelling)
Used the plugin, worked my best to get everything correct.

 * Use markdown syntax for the refs so that they link
I was using a plugin which allows using Latex references. Changed to be independent of the plugin

 * Move from `fipy` to `examples-contributed`
Moved to exemples\flow as per Mr. Guyer suggestion

* Only have the .ipynb file in the commit
The ew commit only have the jupyter notebook as suggested

* I like the streamlines, but maybe clean up the cells at the end with the repetitive looping

wd15 commented:
Hi @Jon83Carvalho, I've worked on your example a little bit, see [the updated notebook](https://github.com/wd15/fipy/blob/fc6ead080feacef7166ddfa5c0ccea5f6812f3e0/examples/flow/stokesNonNewtonian.ipynb).

I did some code formatting and changed the diffusion coefficients so they are updated automatically without having to assign lots of intermediate variables inside the loop. It certainly reduces the number of lines of code. It seems to give the same flow fields (I didn't check the numbers to see if they are exactly the same though). 

One oddity is there wasn't a nice way to do a matrix transpose with fipy which preserves  the variable (lazily). I did it in a round about way. @guyer, has swapaxes. transpose or .T ever worked for rank 2 variables. What have we done for that in the past?

Anyway, @Jon83Carvalho,  I'm happy to merge my changes into this PR assuming you're happy with my changes. There are plenty of other things we could work on to clean it up more, but I'm happy with this as it is after my changes are merged. It doesn't have to be perfect. Good enough is fine.

Jon83Carvalho commented:
> Hi @Jon83Carvalho, I've worked on your example a little bit, see [the updated notebook](https://github.com/wd15/fipy/blob/fc6ead080feacef7166ddfa5c0ccea5f6812f3e0/examples/flow/stokesNonNewtonian.ipynb).
> 
> I did some code formatting and changed the diffusion coefficients so they are updated automatically without having to assign lots of intermediate variables inside the loop. It certainly reduces the number of lines of code. It seems to give the same flow fields (I didn't check the numbers to see if they are exactly the same though).
> 
> One oddity is there wasn't a nice way to do a matrix transpose with fipy which preserves the variable (lazily). I did it in a round about way. @guyer, has swapaxes. transpose or .T ever worked for rank 2 variables. What have we done for that in the past?
> 
> Anyway, @Jon83Carvalho, I'm happy to merge my changes into this PR assuming you're happy with my changes. There are plenty of other things we could work on to clean it up more, but I'm happy with this as it is after my changes are merged. It doesn't have to be perfect. Good enough is fine.

Hi Mr. Wheeler. I am more than glad with the change., Thanks for updating it and giving insights into how to make it a better code for this example. Thus, I have no further comments.

I just want to add a short comment about the streamlines. I removed because when I tried to work on them, I got an error message indicating that the stream viewer requires equally spaced mesh points, and I could only achieve that by generating a similar mesh with the loops.

wd15 commented:
I merged my changes back in. The stream viewer isn't essential.

tkphd commented:
I did some editing for spelling and spacing, and had to adjust `eps=0.05` to `eps=0.025`, otherwise the solver exited after the first step, with no visible change from the initial condition. Filing a PR against @Jon83Carvalho's fork.

Jon83Carvalho commented:
> Very nice contribution @Jon83Carvalho. It's nice to see everything in one notebook, and the example is well written and easy to follow.

Thanks @tkphd for the inputs. I was looking for help to improve the code, so I appreciate your help. I thank @wd15 for improving the text and cleanup/improve code. I am glad that the original Stokes example could be augmented with a different fluid model. I hope it influences others to develop more fluid flow solutions.

guyer commented:
I opened two small pull requests: Jon83Carvalho/fipy#4 and Jon83Carvalho/fipy#5.

@Jon83Carvalho I didn't understand your comment "stream viewer requires equally spaced mesh points". Our `MatplotlibStreamViewer` takes care of gridding irregular data automatically. If it doesn't work for you, though, post the error traceback you get under my [pull request](https://github.com/Jon83Carvalho/fipy/pull/5).

guyer commented on examples/flow/stokesNonNewtonian.ipynb:
```diff
@@ -0,0 +1,498 @@
+   "outputs": [
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAlsAAAJDCAYAAAA8QNGHAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+j8jraAAAgAElEQVR4nOy9eXxURfY2/tze0t3ZO/tCFhKyEELIAgkhkBCSkBUXBAEVRmZUFAUUVBRkRESFcRAQEEEQUHYEQREwIoggENZAIAESIBvZl87W6fRyfn9kbn27ydbOb3zn6/v28/nUpzvpc+vWrVu36txznnOKIyKYYYYZZphhhhlmmPHHQPDfboAZZphhhhl ...
```
@wd15 I realize this is taken directly from [`stokesCavity.py`](https://github.com/usnistgov/fipy/blob/master/examples/flow/stokesCavity.py), but why use `_FaceGradVariable` here? Does `pressure.faceGrad` not work for some reason?

guyer commented:
@wd15:

> has swapaxes. transpose or .T ever worked for rank 2 variables. What have we done for that in the past?

Not that I'm aware of. I think I might have got it working on my `variableMeshes` branch, but that's not much use here.

I think the closest I ever got was for the [elasticity benchmark](https://github.com/guyer/elasticity/blob/master/HuChenElasticity-SteinbachApel.ipynb), but I did all that in reciprocal space:
```python
straing = 0.5j * (nmx.einsum('i...,j...->ij...', ug, g) + nmx.einsum('j...,i...->ij...', ug, g))
```

Jon83Carvalho commented:
> @Jon83Carvalho I like this notebook. After consideration of my [pull](https://github.com/Jon83Carvalho/fipy/pull/4) [requests](https://github.com/Jon83Carvalho/fipy/pull/5), I think this is ready to merge.

Thanks Mr. @guyer. I have no excuse. Originally, I tried hard to put the Stream Viewer, also tried different library stream viewer from Plotly instead of Matplotlib and in both I got a message like "viewer requires equally spaced mesh points" I even reinstalled python and fipy different times to try solve it. But now, I tested your code and also my original code and just worked. No idea what I did wrong before.

guyer commented:
@Jon83Carvalho 

> But now, I tested your code and also my original code and just worked. No idea what I did wrong before.

No worries! If it didn't work for some reason, I wanted to be able to fix it.

wd15 commented:
> @wd15:
> 
> ```python
> straing = 0.5j * (nmx.einsum('i...,j...->ij...', ug, g) + nmx.einsum('j...,i...->ij...', ug, g))
> ```

Transpose made difficult!
